pkg/process/procutil/process_linux.go

```diff
@@ -583,6 +583,40 @@ func (p *Probe) getFDCount(pidPath string) int32 {
 	return int32(len(names))
 }
 
+// getFDCountImproved gets num_fds from /proc/(pid)/fd WITHOUT using the native Readdirnames(),
+// this will skip the step of returning all file names(we don't need) in a dir which takes a lot of memory
+func (p *Probe) getFDCountImproved(pidPath string) int32 {
+	path := filepath.Join(pidPath, "fd")
+
+	if err := p.ensurePathReadable(path); err != nil {
+		return -1
+	}
+
+	d, err := os.Open(path)
+	if err != nil {
+		return -1
+	}
+	defer d.Close()
+
+	b := make([]byte, 8192)
+	count := 0
+
+	for i := 0; ; i++ {
+		n, err := syscall.ReadDirent(int(d.Fd()), b)
+		if err != nil {
+			return -1
+		}
+		if n <= 0 {
```

Comment:
maybe we should leave a comment saying that `n <= 0` when there's no `dirent` structures left to be read on the current directory?

Reply:
This is kinda obvious so I didn't put comment.

Reply:
Can it ever be less than 0?

Reply:
Yes according to the original source code: https://github.com/golang/go/blob/release-branch.go1.14/src/os/dir_unix.go#L57

Reply:
Haha ok, better safe than sorry. lgtm!